src/System.IO.FileSystem/src/System/IO/UnixFileSystem.cs

```diff
@@ -52,24 +52,35 @@ public override void CopyFile(string sourceFullPath, string destFullPath, bool o
 
         public override void MoveFile(string sourceFullPath, string destFullPath)
         {
-            while (Interop.libc.rename(sourceFullPath, destFullPath) < 0)
+            while (Interop.libc.link(sourceFullPath, destFullPath) < 0)
             {
                 int errno = Marshal.GetLastWin32Error();
                 if (errno == Interop.Errors.EINTR) // interrupted; try again
                 {
                     continue;
                 }
-                else if (errno == Interop.Errors.EXDEV) // rename fails across devices / mount points
+                else if (errno == Interop.Errors.EXDEV) // link fails across devices / mount points
                 {
                     CopyFile(sourceFullPath, destFullPath, overwrite: false);
-                    DeleteFile(sourceFullPath);
                     break;
                 }
+                else if (errno == Interop.Errors.ENOENT && !Directory.Exists(Path.GetDirectoryName(destFullPath))) // The destFile parent directory can't be found
```

Comment:
Yes, we should check source as well. I need to add a test case that covers when sourceFullPath has a nonexistent directory in its path like I do with destFullPath and validate that a DirectoryNotFoundException is thrown.